subprojects/core/src/main/java/org/gradle/internal/operations/notify/BuildOperationNotificationBridge.java

```diff
@@ -35,16 +40,45 @@
 
     private Listener operationListener;
     private final BuildOperationListenerManager buildOperationListenerManager;
+    private BuildOperationRecorder buildOperationRecorder;
 
-    BuildOperationNotificationBridge(BuildOperationListenerManager buildOperationListenerManager) {
+    BuildOperationNotificationBridge(BuildOperationListenerManager buildOperationListenerManager, BuildOperationRecorder buildOperationRecorder, GradleInternal gradleInternal) {
         this.buildOperationListenerManager = buildOperationListenerManager;
+        this.buildOperationRecorder = buildOperationRecorder;
+        // ensure we only store events for configuration phase to keep overhead small
+        // when build scan plugin is not applied
+        gradleInternal.rootProject(new Action<Project>() {
+            @Override
+            public void execute(Project project) {
+                project.afterEvaluate(new Action<Project>() {
+                    @Override
+                    public void execute(Project project) {
+                        if (BuildOperationNotificationBridge.this.buildOperationRecorder != null) {
```

Comment:
WDYT about instead of sometimes setting `buildOperationRecorder` to null to signify that we have a build operation listener, we instead use a different method like...

`buildOperationRecorder.stopAndDiscardEvents()`

And in `stopAndDiscardEvents` we could have a debug messages that say something like "X events were discarded" (if there's no listener) and "already stopped, build operation listener such-and-such was added".